compiler/rustc_span/src/hygiene.rs

```diff
@@ -1188,6 +1188,53 @@ impl MacroKind {
     }
 }
 
+/// A set of macro kinds, for macros that can have more than one kind
+#[derive(Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Encodable, Decodable, Hash, Debug)]
+#[derive(HashStable_Generic)]
+pub struct MacroKinds(u8);
+bitflags::bitflags! {
+    impl MacroKinds: u8 {
+        const BANG = 1 << 0;
+        const ATTR = 1 << 1;
+        const DERIVE = 1 << 2;
+    }
+}
+
+impl From<MacroKind> for MacroKinds {
+    fn from(kind: MacroKind) -> Self {
+        match kind {
+            MacroKind::Bang => Self::BANG,
+            MacroKind::Attr => Self::ATTR,
+            MacroKind::Derive => Self::DERIVE,
+        }
+    }
+}
+
+impl MacroKinds {
+    /// Convert the MacroKinds to a static string.
+    ///
+    /// This hardcodes all the possibilities, in order to return a static string.
+    pub fn descr(self) -> &'static str {
+        match self {
+            // FIXME: change this to "function-like macro" and fix all tests
+            Self::BANG => "macro",
+            Self::ATTR => "attribute macro",
+            Self::DERIVE => "derive macro",
+            _ if self == (Self::ATTR | Self::BANG) => "attribute/function macro",
```

Comment:
```suggestion
            const { Self::ATTR | Self::BANG } => "attribute/function macro",
```
should already work, IIRC.

Reply:
@petrochenkov 

Not yet:

```text
error: const blocks cannot be used as patterns
    --> compiler/rustc_span/src/hygiene.rs:1223:19
     |
1223 |             const { Self::ATTR | Self::BANG } => "attribute/function macro",
     |                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^
     |
     = help: use a named `const`-item or an `if`-guard (`x if x == const { ... }`) instead
```

(And even if that were allowed, bitflags' `BitOrd` implementation isn't `[const]`.)

Reply:
(inline const pats where removed: https://github.com/rust-lang/rust/pull/138492)